kernel/genlambda.ml

```diff
@@ -452,6 +459,11 @@ let rec remove_let subst lam =
       if def == def' && body == body' then lam else Llet(id,def',body')
   | _ -> map_lam_with_binders liftn remove_let subst lam
 
+let optimize lam =
+  let lam = simplify lam in
+  let lam = remove_let (subs_id 0) lam in
```

Comment:
I am not too happy about this line. I consider the `remove_let` optimization to be quite dubious already, especially in the presence of primitive arrays (see #12947, #13234). And now, if I understand correctly, `native_compute` will also suffer from it.

Reply:
I can just remove this line if you prefer. My only preoccupation is to generate the same code in native and VM.

Reply:
That said, I stand by my original position in these bug reports, i.e. relying on let-bindings to order computation in Coq is a terrible idea that conflicts with the internal semantics of CIC. The only sane way to force computation order is by pattern matching, i.e. replace `let x := t in u` with `match t with Box x => u end` where `Inductive box (A : Type) := Box : A -> box A`.

Reply:
I can also take a middle-ground mixing the current VM and native heuristic, by only let-substituting values.

Reply:
Most of the values have already been substituted by `simplify` itself. So, the remaining ones are just the ones for creating closures: `Llam`, `Lfix`, and `Lcofix`. That would be fine, since moving them around cannot arbitrarily change the time complexity of an algorithm, not even when taking the garbage collector into account, if I am not mistaken.

And by the way, in my opinion, having the user replace `let x := t in u` with `match Box t with Box x => u end` is just insane.

Reply:
> That would be fine

Patch incoming.

> just insane

Beware, it's not an η expansion as you wrote, to indicate call-by-value you *must* return a Box type, so all strict primitives must have type `Box A` for some `A`, or `unit` if they do not return anything. Indeed, `let _ : unit := t in u` and `match t with tt => u` are two very different things computationally speaking.